File: algorithm/dynamic_programming/src/fibonacci.py
# Dongyoung Kwon @Chuncheonian (ehddud2468@gmail.com)
import collections

# 1. fibonacci-bruteforce
def fib_bruteforce(n: int) -> int:
    if n <= 1:
        return n

    return fib_bruteforce(n-1) + fib_bruteforce(n-2)

# 2. fibonacci-memoization: Top-Down
dp = collections.defaultdict(int)

def fib_memoization(n: int) -> int:
    if n <= 1:
        return n

    if dp[n] == True:
        return dp[n]

    dp[n] = fib_memoization(n-1) + fib_memoization(n-2)
    
    return dp[n]
    
# 3. fibonacci-tabulation: Bottom-Up
dp2 = collections.defaultdict(int)

def fib_tabulation(n: int) -> int:
    dp2[1] = 1

    for i in range(2, n + 1):
        dp2[i] = dp2[i - 1] + dp2[i - 2]

    return dp2[n]

# 4. Space complexity improved fibonacci
def fib_simple(n: int) -> int:
    x, y = 0, 1

    for _ in range(n):
        x, y = y, x + y

    return x

# Testcase
print(fib_bruteforce(5))
print(fib_memoization(5))
print(fib_tabulation(5))
print(fib_simple(5))

File: algorithm/dynamic_programming/src/fibonacci.swift
// Dongyoung Kwon @Chuncheonian (ehddud2468@gmail.com)

// 1. fibonacci - bruteforce
func fibonacciBruteforce(_ n: Int) -> Int {
  n < 2 ? n : fibonacciBruteforce(n - 1) + fibonacciBruteforce(n - 2)
}

// 2. fibonacci - tabulation : Bottom-Up
func fibonacciTabulation(_ n: Int) -> Int {
  var dp: [Int] = [0, 1]

  guard n > 1 else {
    return n
  }

  for i in 2...n {
    dp.append(dp[i - 1] + dp[i - 2])
  }

  return dp[n]
}


// 3. fibonacci - memoization : Top-Down
var dp = [Int: Int]()

func fibonacciMemoization(_ n: Int) -> Int {
  if let v = dp[n] {
    return v
  }

  let newValue = n < 2 ? n : fibonacciMemoization(n - 1) + fibonacciMemoization(n - 2)
  dp[n] = newValue
  return newValue
}


// Testcase
print(fibonacciBruteforce(10))
print(fibonacciTabulation(10))
print(fibonacciMemoization(10))